### PR TITLE
Fix MemoryFile.readAsLines to behave more like File.readAsLines

### PR DESCRIPTION
**Pull request by jamesderlin**
Fix `MemoryFile.readAsLines` to behave more like `File.readAsLines`.
A final newline should not add an empty string as the last element of
the returned `List`.

Additionally make `RecordingFile.readAsLines` always record a final
newline.

Fixes https://github.com/google/file.dart/issues/142 and  https://github.com/google/file.dart/issues/146.

### CODE REVIEW

**Comment by dnfield** [comment]
Can we verify that this doesn't break anything in the Flutter repo? As in, run all the tests from there with this change?

**Comment by jamesderlin** [comment]
I'll test it with the Flutter repo and in google3.  It's certainly possible that this will break some tests (but no more so than using a local file system).

**Comment by dnfield** [comment]
If it breaks anything we should probably just guard it behind a flag so we can opt ourselves into it and not hold up any rolls.

**Comment by jamesderlin** [comment]
In `flutter/packages`:
```
$ rg -l readAsLines | sort | cut -d '/' -f 1-3 | uniq
flutter_tools/lib/src
flutter_tools/test/commands.shard
flutter_tools/test/general.shard
```

So only `flutter_tools` uses `readAsLines`. I modified `flutter_tools/pubspec.yaml` and added:

```yaml
dependency_overrides:
  file:
    path: /MY/PATH/TO/file.dart/packages/file
```
And ran tests.  I hit one failure in `general.shard`:

```
flutter_tools $ ../../bin/cache/dart-sdk/bin/pub run test test/general.shard
03:32 +1249 ~1 -1: test/general.shard/ios/ios_project_migration_test.dart: iOS migration remove framework linking and embedding migration Xcode project is migrated [E]
  Expected: 'keep this 1\n'
              '\t\t\tshellScript = "/bin/sh "$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh\\" embed_and_thin";\n'
              'keep this 2\n'
              '\n'
              ''
    Actual: 'keep this 1\n'
              '\t\t\tshellScript = "/bin/sh "$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh\\" embed_and_thin";\n'
              'keep this 2\n'
              ''
     Which: is different. Both strings start the same, but the actual value is missing the following trailing characters: \n

  package:test_api                                              expect
  test/general.shard/ios/ios_project_migration_test.dart 144:9  main.<fn>.<fn>.<fn>
  dart:async                                                    runZoned
  test/src/common.dart 201:14                                   testWithoutContext.<fn>
  test/src/common.dart 171:18                                   test.<fn>
```

I hit 3 failures in `commands.shard` that are unrelated (they occur without my change; will double-check those later).  I didn't hit any related failures in google3.

> If it breaks anything we should probably just guard it behind a flag so we can opt ourselves into it and not hold up any rolls.

What kind of flag did you have in mind?  Since breakage seems pretty minimal, instead of adding some API to `package:file` (and possibly removing it later), I'm leaning toward adjusting the one failing `flutter_tools` test to tolerate the newline difference.

**Comment by dnfield** [comment]
If it's just one test that can be fixed ahead that's fine

**Comment by jonahwilliams** [comment]
Yeah, one test can be fixed when we pull in the versions. Thank you for fixing this!

**Comment by jamesderlin** [comment]
> Yeah, one test can be fixed when we pull in the versions. Thank you for fixing this!

To be clear: you're okay with fixing the test when it breaks instead of having a forward-compatible in advance, right? (The forward-compatible fix is simple but will involve additional churn.  Meanwhile fixing the breakage would just involve deleting a blank line in the test.)

**Comment by dnfield** [comment]
Do we run the affected test in google3?

If so, we should avoid having any kind of hard break here.

**Comment by jamesderlin** [comment]
We do not run the flutter_tools tests in google3 as part of the global presubmit.  I don't think it's even possible to run those tests manually there.

**Comment by jonahwilliams** [comment]
Yeah, we have no plans to support that. This change LGTM

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
I don't actually have write access, but if @dnfield approves/merges I can upload now and handle rolling into flutter

**Comment by dnfield** [comment]
Done. Strangely, I have write access to this repo but not publish access to the package :)

**Pull request merged**
